refactor: speed up conversion from dataframe to list of records

TColl opened the pull request:
### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Converting big integers to strings was previously done manually by looping over every single value in the DataFrame, after first calling pandas `DataFrame.to_dict()`, which uses `DataFrame.itertuples()` under the hood.

New version drops pandas `to_dict()` in favour of grabbing the data directly and inserting the big integer conversion within that single loop over the full dataset.

This speeds up the overall record collection by a factor of ~2 (primarily because we're only looping through the data once).

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->
N/A

### TEST PLAN
<!--- What steps should be taken to verify the changes -->
Functionality covered by existing unit tests in `dataframe_test.py`

Benchmarking notebook available [here](https://colab.research.google.com/drive/1_b2KKFrscNsbzceocxNpoNHGTS-IHbSL?usp=sharing)

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

TColl committed:
vectorise big integer to string type conversion

TColl commented:
Re. testing; this is my first contribution to this repo and I'm not clear on preferred docstring style. If someone lets me know what you'd like to see I'll happily add docstrings and doctests to cover this refactor!

TColl commented:
For what it's worth, this could be sped up further (at the expense of readability) - Would you prefer to use pandas builtins or is a dict comprehension OK here?

```
def _convert_big_integers(dframe: pd.DataFrame) -> pd.DataFrame:
    return pd.DataFrame(
        {
            col: dframe[col].map(
                lambda v: str(v)
                if isinstance(v, int) and abs(v) > JS_MAX_INTEGER
                else v
            )
            for col in dframe.columns
        }
    )

TColl committed:
add docstrings per contribution guidelines

TColl committed:
drop applymap() for dict comprehension for speedup

ktmud commented:
I'm a huge fan of performance optimization PRs so I did some benchmarking and added one of my own optimizations:

https://colab.research.google.com/drive/1M6wrvyIBIs079IC8RUGCT5I5zOxwOZ33#scrollTo=Fcie22gnZHqr

Based on my tests, the original method is actually the fastest (the result is consistent on both my local machine and Google Collab). Could you share how you tested 3x performance gain with `Series.map`?

TColl commented:
hi @ktmud, thanks for picking this up!

I think I missed something in my original profiling; I've just checked again on my side and I'm seeing the same results you are, across small and large dataframes and varying fractions of large integers. 

I'll admit I'm used to the 'optimisation' I proposed generally being orders of magnitude faster than manual iteration over each value within a dataframe so I was guilty of confirmation bias when I reviewed my original, mistaken benchmarking, but the context of converting the df to a list of dicts and keeping it in that format is the dominant factor here and as such I'm happy to close this out at this stage (unless anyone can think of a better way to handle this - maybe a more performant alternative to pandas `DataFrame.to_dict(orient="records")`?)

ktmud commented:
Let's keep this PR open for a couple of more days and see if people have ideas. The original code had a `TODO: refactor this`, so I'd assume the original author didn't like manual loops, either.

TColl committed:
drop

TColl commented:
I've had another look at this, and dropping pandas `DataFrame.to_dict()` in favour of the underlying operation means we can do the integer conversion during the same single loop over all records in the dataframe, rather than looping over the whole dataframe twice to get to where we need to end up.

At the risk of making myself look stupid, this seems to result in a 2x speedup on my local tests this time round, but I'd appreciate a second pair of eyes on this!

TColl committed:
map integer conversion (readability)

TColl committed:
fix trailing whitespace

TColl commented:
One thing to watch out for with this approach - we lose the [pandas sense-check](https://github.com/pandas-dev/pandas/blob/cb1486c8ef53db474248a0f9b81937b487388431/pandas/core/frame.py#L1568) that all columns have unique names and hence the warning about the resulting dicts not reflecting the source data if they aren't.

It'd be easy enough to add this check into our function here too - either via `warnings.warn()` or through the logger. What would be preferred?

TColl committed:
include negative integers

TColl committed:
grab underlying data directly

TColl commented:
@ktmud I've pushed some further optimisations and added a benchmarking notebook (see edited PR description). Looks like a ~2x speedup on a range of dataframe sizes now 👍

TColl committed:
return list directly

ktmud commented:
> Either way, it'd be easy enough to add this check into our function here too - via warnings.warn() or through the logger. What would be preferred?

Let's use `warnings.warn()` just to be consistent with current behavior.

ktmud reviewed (approved):
I didn't really see the 2x performance gain in my benchmark tests, but the code does look cleaner than before so I'm stamping this.

TColl committed:
raise warning if column names aren't unique

TColl commented:
I've added the check/warning and updated my [benchmarking notebook](https://colab.research.google.com/drive/1_b2KKFrscNsbzceocxNpoNHGTS-IHbSL?usp=sharing) to include the latest code - I'm still seeing a 2x improvement for large dataframes, and more like 5-6x speedup on smaller dataframes. Either way, I think this is good to go now!

I'm surprised you didn't see similar improvements in your tests - could you share what you've tried so I can have a look?

TColl committed:
grab columns once

TColl commented:
I've added the check/warning in and one final tweak (though we're really into diminishing returns at this point). With the latest code I'm seeing 2x speedup on large dataframes and 6-7x speedup on smaller ones - [benchmarking notebook](https://colab.research.google.com/drive/1_b2KKFrscNsbzceocxNpoNHGTS-IHbSL?usp=sharing). I'm surprised you're not seeing similar performance improvements; would you mind sharing how you've tested it?

I think this is finally good to go - thanks for bearing with me after the false start!

ktmud commented:
> I'm surprised you're not seeing similar performance improvements; would you mind sharing how you've tested it?

I just updated the notebook I shared earlier with your code.

ktmud commented:
@TColl I think this looks good now. Congrats on landing your first Superset PR! 🥳

TColl commented:
> > I'm surprised you're not seeing similar performance improvements; would you mind sharing how you've tested it?
> 
> I just updated the notebook I shared earlier with your latest code.

It looks like you're still calling `convert3`, not `convert4` in those benchmarks!

ktmud commented:
> > > I'm surprised you're not seeing similar performance improvements; would you mind sharing how you've tested it?
> > 
> > 
> > I just updated the notebook I shared earlier with your latest code.
> 
> It looks like you're still calling `convert3`, not `convert4` in those benchmarks!

😱 Can't believe I made such a stupid mistake.

The updated results look good now. Thanks!

junlincc commented:
@TColl Thank you so much for your contribution! 
for documentation purpose, @ktmud could you comment here what/ or if there is any risks associate with this refractor? thanks! 

cc @guhan

ktmud commented:
@junlincc It's a simple performance optimization. Risks are captured by passing the existing unit tests, as described in Test Plan.
